### DISCUSSION TITLE
'&' sign in String primary key

### DISCUSSION THREAD

**Discussion by Asky-GH** [yahoo/elide#2331] [Q&A] [answered]
Hi! I have some data in one of my database tables with **id** and **name** columns. Both columns are of type **String**. One of the entries' id contains '**&**' sign in it. And when I'm requesting it by id I'm getting the following response:
```
{
    "errors": [
        {
            "detail": "token recognition error at: &#39;&amp;&#39;"
        }
    ]
}
```
Is there a way to fix such an issue or should I change the id?

**Comment by aklish** [comment, marked as answer]
This is the BNF for parsing a Path string in Elide JSON-API :

https://github.com/yahoo/elide/blob/master/elide-core/src/main/antlr4/com/yahoo/elide/generated/parsers/Core.g4#L45-L56

Basically, you can have alphanumeric and any one of these special characters:

```
MARK
    : '-'
    | '_'
    | '.'
    | '!'
    | '~'
    | '='  //For BASE64 IDs
    | '%'  //For URL encoded IDs
    | '*'
    | '\''
    | '('
    | ')'
```

**Reply by Asky-GH** [reply]
@aklish, is '**&**' not included in the list above because it's being used in other places of the url (e.g. **/books?include=...&filter=...**)?

**Reply by aklish** [reply]
I believe that was the intent, but I think we can relax that restriction.  Interested in doing a small PR to patch?

**Reply by Asky-GH** [reply]
I’d be happy but I’m not sure what exactly should be done. Is it as simple as adding the ‘&’ sign to the Core.g4 file or maybe some further changes should be made? Won’t it break existing functionality?

**Comment by aklish** [comment]
There are two options:
1. Change the parser and relax what can constitute a path segment.  This just involves change the parser file above and adding a unit test to https://github.com/yahoo/elide/blob/master/elide-core/src/test/java/com/yahoo/elide/endpoints/ResourceTest.java#L19.  This shouldn't break anything as we are relaxing what is permissible.  
2. Second option is slightly more involved, but we could create a new data type (UrlEncodedString) that stores a String inside of it.  We can create a [Serde](https://elide.io/pages/guide/v6/09-clientapis.html#user-type-registration) for this type that converts to a Java String by url encoding/decoding the contents.  Any Elide model that wants to use a String as an ID field can then use this type instead.  The client then simply has to url encode the String properly before sending it to the service.

The second option doesn't require an Elide change but would probably be useful for others.

**Reply by Asky-GH** [reply]
Hi @aklish!
Sorry for such a late response (was too busy at work).
If you don't mind, I'd go with the first option (as already mentioned above I'm short on time recently).

**Reply by aklish** [reply]
Another customer asked for this same week so i ended up doing a small PR on mainline:

https://github.com/yahoo/elide/pull/2342

If you want this on Elide 4 or 5, you can take a diff of the changes and submit a new PR against the elide-4.x or elide-5.x branches respectively.

**Reply by Asky-GH** [reply]
Since I'm using Elide 5, submitted a new PR against elide-5.x branch (#2345)

**Comment by aklish** [comment]
Released in:

https://github.com/yahoo/elide/releases/tag/5.1.0
